### PR TITLE
Feature/gasvm refactoring/genetic algorithm factory

### PR DESCRIPTION
**Pull request by Regis128**
Added minimal and maximal fillup parameters, few exceptions, GeneticAlgorithmFactory and module dependency changes

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -13,7 +13,7 @@ environment:
 configuration:
   - Debug
   - Release
-version: 3.0.0.{build}
+version: 3.4.0.{build}
```
Comment (gmrukwa) [inline appveyor.yml]: At last someone remembered! :D

File: src/Spectre.libGenetic.Tests/CrossoverOperatorTest.cpp [hunk 2]
```diff
@@ -153,4 +161,15 @@ TEST_F(CrossoverOperatorTest, cuts_are_from_uniform_distribution)
         EXPECT_LT(count, meanCount + allowedCountMiss) << i;
     }
 }
+
+TEST_F(CrossoverOperatorTest, test_if_return_first_individual_if_fillup_is_too_low)
```
Comment (gmrukwa) [inline src/Spectre.libGenetic.Tests/CrossoverOperatorTest.cpp]: Missing opposite case with assumptions against situation, when fillup is too high.

File: src/Spectre.libGenetic.Tests/MutationOperatorTest.cpp [hunk 3]
```diff
@@ -167,16 +167,31 @@ TEST_F(MutationTest, toggles_in_approximate_percentage_of_cases_for_specified_bi
...
+        for (size_t j = 0u; j < individual.size(); ++j)
         {
             numberOfToggles += last[j] != individual[j];
         }
     }
     EXPECT_LT(numberOfToggles, expectedNumberOfToggles + allowedMissCount);
     EXPECT_GT(numberOfToggles, expectedNumberOfToggles - allowedMissCount);
 }
+
+TEST_F(MutationTest, test_if_returns_original_if_fillup_is_too_low)
```
Comment (gmrukwa) [inline src/Spectre.libGenetic.Tests/MutationOperatorTest.cpp]: This test is a little bit confusing. If `individual` is all true, please, name it so, since otherwise context is lost.

On the other hand, why do we need multiple trials here? If we set `BIT_SWAP_RATE` to `ALWAYS` (or `NEVER`, dependent on the content of `individual`), we can test it once - this behaviour is deterministic, right?

File: src/Spectre.libGenetic/CrossoverOperator.cpp [hunk 4]
```diff
@@ -17,14 +17,25 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+#include <random>
+#include "Spectre.libStatistics/Statistics.h"
```
Comment (gmrukwa) [inline src/Spectre.libGenetic/CrossoverOperator.cpp]: Why the hell do we need it here? O.o

File: src/Spectre.libGenetic/InconsistentMinimalAndMaximalFillupException.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+#include "Spectre.libException/InconsistentArgumentSizesException.h"
+
+namespace spectre::algorithm::genetic
+{
+/// <summary>
+/// Thrown when minimal and maximal fillup values are inconsistent.
+/// </summary>
+class InconsistentMinimalAndMaximalFillupException :
+    public core::exception::InconsistentArgumentSizesException
```
Comment (gmrukwa) [inline src/Spectre.libGenetic/InconsistentMinimalAndMaximalFillupException.h]: This exception is not intended for such a case. It is for collections that should be of equal size. I'd suggest using `ArgumentOutOfRangeException` rather than this.

File: src/Spectre.libGenetic/Spectre.libGenetic.vcxproj [hunk 6]
```diff
@@ -1,5 +1,6 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\packages\opencv3.3.0.0.1\build\native\opencv3.props" Condition="Exists('..\packages\opencv3.3.0.0.1\build\native\opencv3.props')" />
```
Comment (gmrukwa) [inline src/Spectre.libGenetic/Spectre.libGenetic.vcxproj]: Why the hell does `libGenetic` require `OpenCV`?

File: src/Spectre.libGenetic/packages.config [hunk 7]
```diff
@@ -1,4 +1,6 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <packages>
   <package id="Microsoft.Gsl" version="0.1.2.1" targetFramework="native" />
+  <package id="opencv3" version="3.0.0.1" targetFramework="native" />
```
Comment (gmrukwa) [inline src/Spectre.libGenetic/packages.config]: No. No. NO! :D

File: src/Spectre.libStatistics.Tests/DifferentiatingFeaturesEstimatorTest.cpp [hunk 8]
```diff
@@ -92,13 +92,15 @@ TEST(DifferentiatingFeaturesEstimator, throws_for_inconsistent_number_of_feature
     EXPECT_THROW(estimator.Estimate(narrowDataset, secondDataset), InconsistentNumberOfFeaturesException);
 }
 
+/* disabled because it fails sometimes and needs to be checked
```
Comment (gmrukwa) [inline src/Spectre.libStatistics.Tests/DifferentiatingFeaturesEstimatorTest.cpp]: This way we will forget it as soon, as this gets merged. Check [this](https://stackoverflow.com/a/7208119/3067956). I think that gtest will at least keep spam with messages like: *Hey! You've got 1 disabled test!*.

File: src/Spectre.libGenetic.Tests/CrossoverOperatorTest.cpp (hunk 2 above)
Comment (Regis128) [reply src/Spectre.libGenetic.Tests/CrossoverOperatorTest.cpp]: done

File: src/Spectre.libGenetic.Tests/MutationOperatorTest.cpp (hunk 3 above)
Comment (Regis128) [reply src/Spectre.libGenetic.Tests/MutationOperatorTest.cpp]: done

File: src/Spectre.libGenetic/CrossoverOperator.cpp (hunk 4 above)
Comment (Regis128) [reply src/Spectre.libGenetic/CrossoverOperator.cpp]: done

File: src/Spectre.libGenetic/InconsistentMinimalAndMaximalFillupException.h (hunk 5 above)
Comment (Regis128) [reply src/Spectre.libGenetic/InconsistentMinimalAndMaximalFillupException.h]: done

File: src/Spectre.libGenetic/Spectre.libGenetic.vcxproj (hunk 6 above)
Comment (Regis128) [reply src/Spectre.libGenetic/Spectre.libGenetic.vcxproj]: done

File: src/Spectre.libGenetic/packages.config (hunk 7 above)
Comment (Regis128) [reply src/Spectre.libGenetic/packages.config]: done

File: src/Spectre.libStatistics.Tests/DifferentiatingFeaturesEstimatorTest.cpp (hunk 8 above)
Comment (Regis128) [reply src/Spectre.libStatistics.Tests/DifferentiatingFeaturesEstimatorTest.cpp]: done

File: src/Spectre.libGenetic/InconsistentMinimalAndMaximalFillupException.h (hunk 5 above)
Comment (Regis128) [reply src/Spectre.libGenetic/InconsistentMinimalAndMaximalFillupException.h]: ArgumentOutOfRangeException is also not exactly this error. I will create new error for this purpose from BaseException

File: src/Spectre.libGenetic.Tests/CrossoverOperatorTest.cpp [hunk 9]
```diff
@@ -41,25 +49,25 @@ class CrossoverOperatorTest: public ::testing::Test
     const Seed SEED = 0;
     const Individual true_individual = Individual({ true, true, true, true });
     const Individual false_individual = Individual({ false, false, false, false });
-    CrossoverOperator crossover;
+    std::unique_ptr<CrossoverOperator> crossover;
```
Comment (Sand3r-) [inline src/Spectre.libGenetic.Tests/CrossoverOperatorTest.cpp]: What is the reason behind this change?

File: src/Spectre.libGenetic.Tests/CrossoverOperatorTest.cpp [hunk 10]
```diff
@@ -41,25 +49,25 @@ class CrossoverOperatorTest: public ::testing::Test
...
+        crossover = std::make_unique<CrossoverOperator>(SEED);
     }
 };
 
 TEST_F(CrossoverOperatorTest, throws_on_inconsistent_sizes)
 {
     const Individual shorter({ false });
     const Individual longer({ false, true });
-    EXPECT_THROW(crossover(shorter, longer), InconsistentChromosomeLengthException);
+    EXPECT_THROW(crossover->operator()(shorter, longer), InconsistentChromosomeLengthException);
```
Comment (Sand3r-) [inline src/Spectre.libGenetic.Tests/CrossoverOperatorTest.cpp]: Do whatever you feel is appropriate, but personally I believe that ```crossover->operator()(shorter, longer)``` looks a bit hacky.
Why not to try ```(*crossover.get())(shorter, longer)```? It's a bit shorter and to me it makes much more sense to extract the pointer, dereference it and perform the call.

File: src/Spectre.libGenetic/CrossoverOperator.cpp [hunk 11]
```diff
@@ -44,7 +53,23 @@ Individual CrossoverOperator::operator()(const Individual &first, const Individu
     phenotype.reserve(first.size());
     phenotype.insert(phenotype.end(), first.begin(), endOfFirst);
     phenotype.insert(phenotype.end(), beginningOfSecond, second.end());
-    Individual individual(std::move(phenotype));
-    return std::move(individual);
+
+    auto fillup = 0u;
+    for(auto bit: phenotype)
+    {
+        if(bit)
```
Comment (Sand3r-) [inline src/Spectre.libGenetic/CrossoverOperator.cpp]: It can be optimized to avoid branching:
```
for(auto bit: phenotype)
{
    fillup += bit;
}
```

File: src/Spectre.libGenetic/GeneticAlgorithmFactory.cpp [hunk 12]
```diff
@@ -0,0 +1,61 @@
+/*
+* GeneticAlgorithmFactory.cpp
+* Builds different versions of GeneticAlgorithm
+*
+Copyright 2017 Spectre Team
```
Comment (Sand3r-) [inline src/Spectre.libGenetic/GeneticAlgorithmFactory.cpp]: 2018 :wink:

File: src/Spectre.libGenetic/GeneticAlgorithmFactory.h [hunk 13]
```diff
@@ -0,0 +1,87 @@
+/*
+* GeneticAlgorithmFactory.h
+* Builds different versions of GeneticAlgorithm
+*
+Copyright 2017 Spectre Team
```
Comment (Sand3r-) [inline src/Spectre.libGenetic/GeneticAlgorithmFactory.h]: ...And here, as well and in later ones.

File: src/Spectre.libGenetic/MutationOperator.cpp [hunk 14]
```diff
@@ -52,6 +61,23 @@ Individual MutationOperator::operator()(Individual &&individual)
...
-    return individual;
+
+    auto fillup = 0u;
+    for (auto bit : individual)
+    {
+        if (bit)
+        {
+            ++fillup;
+        }
+    }
```
Comment (Sand3r-) [inline src/Spectre.libGenetic/MutationOperator.cpp]: Same optimization as mentioned previously can be applied here.

File: src/Spectre.libGenetic/Scorer.cpp [hunk 15]
```diff
@@ -18,29 +18,47 @@ limitations under the License.
...
+    }
+    else
+    {
+        const auto optimalChunksNumber = 1;
+        const auto populationSize = static_cast<int>(generation.size());
+        const auto firstIndividual = generation.begin();
+        #pragma omp parallel for schedule(dynamic, optimalChunksNumber) num_threads (m_NumberOfCores)
+        for (auto i = 0; i < populationSize; ++i)
+        {
+            const auto& individual = *(firstIndividual + i);
```
Comment (Sand3r-) [inline src/Spectre.libGenetic/Scorer.cpp]: Any particular reason to use ```*(firstIndividual + i)``` instead of ```generation[i]```?

File: src/Spectre.libGenetic.Tests/CrossoverOperatorTest.cpp (hunk 9 above)
Comment (Regis128) [reply src/Spectre.libGenetic.Tests/CrossoverOperatorTest.cpp]: Just it was easier for me to create pointer in setUp function, and to be more consistent with other tests

File: src/Spectre.libGenetic.Tests/CrossoverOperatorTest.cpp (hunk 10 above)
Comment (Regis128) [reply src/Spectre.libGenetic.Tests/CrossoverOperatorTest.cpp]: Maybe you're right, personally don't mind it ;], so I changed it. Done

File: src/Spectre.libGenetic/CrossoverOperator.cpp (hunk 11 above)
Comment (Regis128) [reply src/Spectre.libGenetic/CrossoverOperator.cpp]: Good point, done

File: src/Spectre.libGenetic/GeneticAlgorithmFactory.cpp (hunk 12 above)
Comment (Regis128) [reply src/Spectre.libGenetic/GeneticAlgorithmFactory.cpp]: Done

File: src/Spectre.libGenetic/GeneticAlgorithmFactory.h (hunk 13 above)
Comment (Regis128) [reply src/Spectre.libGenetic/GeneticAlgorithmFactory.h]: Done

File: src/Spectre.libGenetic/MutationOperator.cpp (hunk 14 above)
Comment (Regis128) [reply src/Spectre.libGenetic/MutationOperator.cpp]: Done

File: src/Spectre.libGenetic/Scorer.cpp (hunk 15 above)
Comment (Regis128) [reply src/Spectre.libGenetic/Scorer.cpp]: Honestly I wasn't sure at first why is it like that. And I still can't see why, so changed ;]

File: src/Spectre.libGenetic.Tests/CrossoverOperatorTest.cpp (hunk 9 above)
Comment (gmrukwa) [reply src/Spectre.libGenetic.Tests/CrossoverOperatorTest.cpp]: During the tests we need to recreate this object. It's much easier to go for `unique_ptr`.

File: src/Spectre.libGenetic/InconsistentMinimalAndMaximalFillupException.h (hunk 5 above)
Comment (gmrukwa) [reply src/Spectre.libGenetic/InconsistentMinimalAndMaximalFillupException.h]: Good idea.

File: src/Spectre.libGenetic/InconsistentMinimalAndMaximalFillupException.h [hunk 16]
```diff
@@ -18,23 +18,23 @@ limitations under the License.
...
 {
 public:
     /// <summary>
     /// Initializes a new instance of the <see cref="InconsistentMinimalAndMaximalFillupException"/> class.
     /// </summary>
     /// <param name="minimal">The minimal fillup value.</param>
     /// <param name="maximal">The maximal fillup value.</param>
-    InconsistentMinimalAndMaximalFillupException(size_t minimal, size_t maximal);
+    InconsistentMinimalAndMaximalFillupException(size_t minimal, size_t maximal):
+        ExceptionBase("minimal value of fillup: " + std::to_string(minimal) + " is greater than its maximum value: " + std::to_string(maximal)) { }
```
Comment (gmrukwa) [inline src/Spectre.libGenetic/InconsistentMinimalAndMaximalFillupException.h]: Implementation should be in `.cpp` file.

File: src/Spectre.libGenetic/packages.config (hunk 7 above)
Comment (gmrukwa) [reply src/Spectre.libGenetic/packages.config]: Thanks :)

**Comment by gmrukwa** [comment]
[Merge updates of GA](https://trello.com/c/rHY973Dc/344-merge-updates-of-ga)
